### PR TITLE
Nested timeouts

### PR DESCRIPTION
**Pull request by stevelinton**
This is built on top of PR #331 and includes a first try and support for nesting alarms (without the break loop being involved).  So you can do CallWithTimeout inside another one and have the inner one go off and be dealt with while the outer one still [appears to] run on.  The code is a little complex, so definitely WIP for now.

### CODE REVIEW

**Comment by stevelinton** [comment]
Now rebased onto master because PR #331 has been merged.

As far as I can tell this works, but the code to restore the outer timeout after the inner one completes
is a bit delicate, so I'd welcome review or testing, but this could be merged.

File: src/gap.c [hunk 1]
```diff
@@ -1302,41 +1294,16 @@ Obj FuncCALL_WITH_CATCH( Obj self, Obj func, Obj args )
...
-                                  ELM_PLIST(plain_args,4), ELM_PLIST(plain_args,5));
-        break;
-      case 6: result = CALL_6ARGS(func, ELM_PLIST(plain_args,1),
-                                  ELM_PLIST(plain_args,2), ELM_PLIST(plain_args,3),
-                                  ELM_PLIST(plain_args,4), ELM_PLIST(plain_args,5),
-                                  ELM_PLIST(plain_args,6));
-        break;
-      default: result = CALL_XARGS(func, plain_args);
+    result = CallFuncList(func, args);
+    SET_ELM_PLIST(res,1,True);
```
Comment (fingolfin) [inline src/gap.c]: The indention here is wrong (at least two more spaces are needed, as this is in an `else` block). This causes several unnecessary diffs later on.

Off-topic: One of these days, I'd love if we could agree on a common code formatting standard for the kernel ;).

File: src/gap.c [hunk 2]
```diff
@@ -1278,13 +1277,6 @@ Obj FuncCALL_WITH_CATCH( Obj self, Obj func, Obj args )
       ErrorMayQuit("CALL_WITH_CATCH(<func>,<args>): <func> must be a function",0,0);
     if (!IS_LIST(args))
       ErrorMayQuit("CALL_WITH_CATCH(<func>,<args>): <args> must be a list",0,0);
-    if (!IS_PLIST(args))
-      {
-        plain_args = SHALLOW_COPY_OBJ(args);
-        PLAIN_LIST(plain_args);
-      }
-    else 
-      plain_args = args;
```
Comment (fingolfin) [inline src/gap.c]: So why was the above code removed? This change is not mentioned in the commit message. Why is it safe? (I am not saying it is, but it is surpring to me to find it mingled with the rest).

File: src/calls.c [hunk 3]
```diff
@@ -1543,12 +1533,27 @@ Obj FuncCALL_FUNC_LIST (
...
+{
+  /* check that the second argument is a list                            */
+    while ( ! IS_SMALL_LIST( list ) ) {
+        list = ErrorReturnObj(
+            "CallFuncList: <list> must be a small list",
+            0L, 0L,
+            "you can replace <list> via 'return <list>;'" );
+    }
+    return CallFuncList(func, list);
+}
```
Comment (fingolfin) [inline src/calls.c]: Breaking out `CallFuncList` from `FuncCALL_FUNC_LIST` seems like a very good idea, though it would of course be nicer if this was done in a separate commit, instead of mingling it with the change that is actually described in the commit message. At the least, I'd expect the commit message to mention this change.

File: src/gap.c [hunk 4]
```diff
@@ -1374,75 +1341,91 @@ Obj FuncSetUserHasQuit( Obj Self, Obj value)
...
+    Int iseconds, imicroseconds;
+    Int curr_seconds, curr_microseconds, curr_nanoseconds;
+    Int restore_seconds = 0, restore_microseconds = 0;
+      
     if (!SyHaveAlarms)
       ErrorMayQuit("CALL_WITH_TIMEOUT: timeouts not supported on this system", 0L, 0L);
     if (!IS_INTOBJ(seconds) || 0 > INT_INTOBJ(seconds))
-      ErrorMayQuit("CALL_WITH_TIMEOUT(<seconds>, <microseconds>, <func>,<args>): <seconds> must be a non-negative small integer",0,0);
+      ErrorMayQuit("CALL_WITH_TIMEOUT(<seconds>, <microseconds>, <func>,<args>):"
+		   " <seconds> must be a non-negative small integer",0,0);
```
Comment (fingolfin) [inline src/gap.c]: This seems like a purely cosmetical change; well, if you do that already, perhaps insert a space in `<func>,<args>`?

File: src/gap.c [hunk 5]
```diff
@@ -1374,75 +1341,91 @@ Obj FuncSetUserHasQuit( Obj Self, Obj value)
...
+	  /* The existing timeout will go off before we would so don't bother to set a timeout
+	   just reset the existing one and call the function */
+	  iseconds = curr_seconds;
+	  imicroseconds = curr_microseconds;
+	  newJumpBuf = 0;
+	} else {
+	/* We would time out before the other function, so we need to
+	   set our timeout, but remember to reset things later */
+	mayNeedToRestore = 1;
+	newJumpBuf = 1;
```
Comment (fingolfin) [inline src/gap.c]: Isn't the indention level for this last `else` block wrong again? It looks like that in the github interface, at least...

File: src/gap.c [hunk 6]
```diff
@@ -1374,75 +1341,91 @@ Obj FuncSetUserHasQuit( Obj Self, Obj value)
...
-                                  ELM_PLIST(plain_args,2), ELM_PLIST(plain_args,3),
-                                  ELM_PLIST(plain_args,4), ELM_PLIST(plain_args,5));
-        break;
-      case 6: result = CALL_6ARGS(func, ELM_PLIST(plain_args,1),
-                                  ELM_PLIST(plain_args,2), ELM_PLIST(plain_args,3),
-                                  ELM_PLIST(plain_args,4), ELM_PLIST(plain_args,5),
-                                  ELM_PLIST(plain_args,6));
-        break;
-      default: result = CALL_XARGS(func, plain_args);
+    }
```
Comment (fingolfin) [inline src/gap.c]: Hrm, it's really a bit difficult to see what changed regarding the timer code, with the `CallFuncList` change mixed in. So, yeah, having `CallFuncList` added in a separate prior commit would be really helpful for reviewers.

File: src/calls.c (hunk 3 above)
Comment (stevelinton) [inline src/calls.c]: You're right of course. I need to get out of the habit to doing tidying and minor refactoring as I "run into them" rather than as separate commits. For now I'll adjust the commit message.

File: src/gap.c (hunk 2 above)
Comment (stevelinton) [inline src/gap.c]: Part of the refactoring of CallFuncList. That's where this code belongs.

File: src/calls.c (hunk 3 above)
Comment (fingolfin) [inline src/calls.c]: The great thing about git, if you use it right, is that you do not need to get out of that habit. Just use `git add -p` to selectively add only parts of your changes (even to individual files) to a commit. I do that all the time. Then, when I am done, I use `git rebase -i` to sort the various I made, fold them together as appropriate, reorder, etc.

This can even be done retroactively, to cleanup and split this commit, though of course it is now more work, since the same piece of code (in `FuncCALL_WITH_TIMEOUT`) would get changed twice.

**Comment by stevelinton** [comment]
OK. Tome for "Adventures in GIT" .... I'll try and split the commit.

**Comment by fingolfin** [comment]
This looks pretty good to me now, thanks. Though I'd of course prefer if the `fixup!` was squashed into its parent. :)

**Comment by alex-konovalov** [comment]
Shall we merge this then? May we have some tests added to timeout.tst to exercise nested timeouts?

**Comment by stevelinton** [comment]
I’ll add some tests — might take a little thought.

**Comment by stevelinton** [comment]
I reran the travis test which had failed, and it passed. No idea what was going on.

**Comment by alex-konovalov** [comment]
Maybe times should be adjusted even further - I've run it twice in Travis, first time it failed, but passed in the 2nd run.

**Comment by alex-konovalov** [comment]
We should remember to update documentation in lib/function.gd which at the moment says that timeouts may not be nested.

**Comment by fingolfin** [comment]
Thanks for the rebase.

Just to document the status of this: As far as I understand, this PR looks basically good, but there is some concern about timeout signals possibly being lost, causing occassional test failures.

But the refactoring commit is a no-brainer, so I'll cherry pick it right away.

**Comment by fingolfin** [comment]
When I manually run the tests, I get some diffs... Huh? Observe:
```
gap> Test("tst/testinstall/timeout.tst");
########> Diff in tst/testinstall/timeout.tst, line 25:
# Input is:
CallWithTimeout(100000, spin2,200,50000);
# Expected output:
[ true, [ false ] ]
# But found:
[ true, [ true ] ]
########
########> Diff in tst/testinstall/timeout.tst, line 31:
# Input is:
CallWithTimeout(10000, spin2,50,200000);
# Expected output:
[ false ]
# But found:
[ true, [ true ] ]
########
########> Diff in tst/testinstall/timeout.tst, line 33:
# Input is:
ct := 0;; CallWithTimeout(200000, function() while true do spin2(50,500000); ct := ct+1; od; end);; 2 <= ct; ct <= 10;
# Expected output:
true
true
# But found:
true
false
########
timeout.tst
GAP4stones: 222
false
```

**Comment by alex-konovalov** [comment]
With the following patch
```
--- a/tst/testinstall/timeout.tst
+++ b/tst/testinstall/timeout.tst
@@ -22,15 +22,15 @@ gap> CallWithTimeoutList(50000,spinFor,[1]);
 [ true ]
 gap> spin2 := function(use, timeout) return CallWithTimeout(timeout, spinFor, use); end;
 function( use, timeout ) ... end
-gap> CallWithTimeout(100000, spin2,200,50000);
+gap> CallWithTimeout(100000, spin2,2000,50000);
 [ true, [ false ] ]
 gap> CallWithTimeout(100000, spin2,50,200000);
 [ true, [ true ] ]
 gap> CallWithTimeout(1000000, spin2,50,200000);
 [ true, [ true ] ]
-gap> CallWithTimeout(10000, spin2,50,200000);                                           
+gap> CallWithTimeout(10000, spin2,500,200000);
 [ false ]
-gap> ct := 0;; CallWithTimeout(200000, function() while true do spin2(50,500000); ct := ct+1; od; end);; 2 <= ct; ct <= 10;
+gap> ct := 0;; CallWithTimeout(500000, function() while true do spin2(500,500000); ct := ct+1; od; end);; 2 <= ct; ct <= 10;
 true
 true
 gap> STOP_TEST( "timeout.tst", 330000);
```
It takes a couple of seconds longer, but now works fine in both runs from `make testinstall` and also when started as `bin/gap.sh tst/testinstall.g`. @stevelinton could you please update this PR so we can merge it? Also, documentation in lib/function.gd should not say that timeouts may not be nested then.

**Comment by fingolfin** [comment]
Hmm, I am not quite sure I understand -- are you saying you understand why it previously sometimes worked, and sometimes (when re-run) did not work? If so, could you explain that, and how your propose change fixes it?

**Comment by alex-konovalov** [comment]
I think that because timings are given in microseconds, small timeouts are too tiny to be deterministic.  (see @rbehrends comments in https://github.com/gap-system/gap/pull/331). I've just tweaked them to what seems to be more reproducible. 

Without the patch, `make testinstall` was passing without packages, but failing with same three diffs as shown by @fingolfin above - but I am pretty sure that packages are not interfering with this.

**Comment by markuspf** [comment]
Is there anything in the way of merging this for 4.8.3? Otherwise, we should move it to the 4.9 milestone and merge it.

**Comment by alex-konovalov** [comment]
I suggest to apply the following patch which makes the test more robust, and then merge it:
```
--- a/tst/testinstall/timeout.tst
+++ b/tst/testinstall/timeout.tst
@@ -22,15 +22,15 @@ gap> CallWithTimeoutList(50000,spinFor,[1]);
 [ true ]
 gap> spin2 := function(use, timeout) return CallWithTimeout(timeout, spinFor, use); end;
 function( use, timeout ) ... end
-gap> CallWithTimeout(100000, spin2,200,50000);
+gap> CallWithTimeout(100000, spin2,2000,50000);
 [ true, [ false ] ]
 gap> CallWithTimeout(100000, spin2,50,200000);
 [ true, [ true ] ]
 gap> CallWithTimeout(1000000, spin2,50,200000);
 [ true, [ true ] ]
-gap> CallWithTimeout(10000, spin2,50,200000);                                           
+gap> CallWithTimeout(10000, spin2,500,200000);
 [ false ]
-gap> ct := 0;; CallWithTimeout(200000, function() while true do spin2(50,500000); ct := ct+1; od; end);; 2 <= ct; ct <= 10;
+gap> ct := 0;; CallWithTimeout(500000, function() while true do spin2(500,500000); ct := ct+1; od; end);; 2 <= ct; ct <= 10;
 true
 true
 gap> STOP_TEST( "timeout.tst", 330000);
```
It takes a couple of seconds longer, but now works fine in both runs from `make testinstall` and also when started as `bin/gap.sh tst/testinstall.g`. Also, documentation in lib/function.gd should not say that timeouts may not be nested then.

**Comment by alex-konovalov** [comment]
@stevelinton after the merge, there is a fresh error when it's compiled on yin with -Werror:
```
gcc -I. -DCONFIG_H  -I/mnt/disk2/hudson-slave/workspace/GAP-dev-compilers/GAPCOPTS/64build/GAPREADLINE/readline/label/64bit/GAP-git-snapshot/bin/x86_64-pc-linux-gnu-gcc-default64/extern/gmp/include -Wall -g -O2 -m64 -m64 -g -O2 -Werror -o gap.o -c ../../src/gap.c
cc1: warnings being treated as errors
../../src/gap.c: In function FuncCALL_WITH_TIMEOUT:
../../src/gap.c:1355: warning: curr_microseconds may be used uninitialized in this function
make[1]: *** [gap.o] Error 1
make[1]: Leaving directory `/mnt/disk2/hudson-slave/workspace/GAP-dev-compilers/GAPCOPTS/64build/GAPREADLINE/readline/label/64bit/GAP-git-snapshot/bin/x86_64-pc-linux-gnu-gcc-default64'
make: *** [compile] Error 2
Build step 'Execute shell' marked build as failure
```

**Comment by alex-konovalov** [comment]
This was reverted in PR #668 so nested timeouts will not be available in GAP 4.8.3
